File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function(nums) {
    if (!nums.length) return 0;
    
    const visited = new Set(nums)
    const subs = {}
    let max = -Infinity
    
    // find the start of each subsequence
    for (const num of nums) {
        if (!visited.has(num - 1)) {
            subs[num] = 1
        }
    }
        
    // for each subsequence, lookup set for consecutives
    for (const sub in subs) {
        let consecutive = Number(sub) + 1
        let count = 1
        
        while (visited.has(consecutive)) {
            // subs[sub]++
            consecutive++
            count++
        }
        
        max = Math.max(max, count)
    }
    
    return max;
};

File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # nums_set for O(1) lookup
        nums_set = set(nums)
        subsets = {}
        max_subset = 0
        
        # find start of each subset, initialize subset size to 1
        for num in nums:
            if (num - 1) not in nums_set:
                subsets[num] = 1
        
        # for each start, increment its value for each exisitng consecutive
        for subset in subsets:
            consecutive = subset + 1
            
            while consecutive in nums_set:
                subsets[subset] += 1
                consecutive += 1
                
            max_subset = max(max_subset, subsets[subset])
            
        return max_subset